---
title: "Introduction to Bioconductor"
author: "Miao Yu"
date: "2015-05-05"
output: html_document
---

## RNA-seq data analysis

```{r}
lengths = c(100,200,300,100,100)
mat = cbind(c(1,1,0,1,0),c(1,1,1,1,1),c(0,1,1,0,1))
lengths %*% mat
theta.hat = c(1, 2, 3) / 10000
w = 1000
mat %*% theta.hat * lengths * w

counts = c(125,350,300,125,100)
LHS = counts/(lengths * w)
lm.fit(mat, LHS)$coefficients
counts2 = c(60,320,420,60,140)
LHS2 = counts2/(lengths * w)
lm.fit(mat, LHS2)$coefficients

genes <- read.table("data/SRR1039508.genes.results", header=TRUE)
isoforms <- read.table("data/SRR1039508.isoforms.results", header=TRUE)

library(dplyr)
data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAASCAYAAABb0P4QAAABCklEQVR42u3Tv2rCUBQGcCkFF11cXCooDoEo4tAgaZdMhTyIu2PBQRydBBfXTtIX8IF8iOTcvzk9J8S0EBsSHerQCx/c4d4f383NbTT+x3L3ie/bQ2nmq32aMAybpdhi84Hr4wljiAsBwQEUUqBUMl3HaCnICy6CBIFkTKIiTGt9HciNAL5bSaVQGY3amPogZKAQhHEzrQjSaKxBa209kFsBt8qa8RENtbKUhDBMknpg1fwtqPi7naN0GgHiNjCf8+1S+LKqgA+u67aGY6/3K5g1jeOoAD6N/E4QBI+55jiv7b7rTQfj2dtP0Bibz/nX4URREeR9lO7FhsOp/3J+p1UymDw7hYZ3P74A6xEjKDEtv3EAAAAASUVORK5CYII=isoforms %>% group_by(gene_id) %>% summarize(sum = sum(FPKM))  

genes2 <- genes[genes$FPKM > 0,]
genes2$gene_id <- droplevels(genes2$gene_id)
isoforms2 <- isoforms[isoforms$gene_id %in% genes2$gene_id,]
isoforms2$gene_id <- droplevels(isoforms2$gene_id)
stopifnot(all(genes2$gene_id == levels(isoforms2$gene_id)))
median(genes2$FPKM)
mean(table(isoforms2$gene_id,isoforms2$IsoPct>95)[,2])
max.iso <- table(isoforms2$gene_id,isoforms2$IsoPct>95)[,2]
boxplot(split(log10(genes2$FPKM), cut(max.iso, 5)), xlab="max.iso", ylab="log10 FPKM")

num.iso <- as.numeric(table(isoforms2$gene_id))
plot(num.iso, max.iso)
barplot(table(num.iso))
barplot(table(num.iso[num.iso < 15]))

library(pasillaBamSubset)
bam.file <- untreated3_chr4()
library(Rsamtools)
bf <- BamFile(bam.file)
library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)
txdb <- TxDb.Dmelanogaster.UCSC.dm3.ensGene
ebg <- exonsBy(txdb, by="gene")

chr4.idx <- all(seqnames(ebg) == "chr4")
ebg.sub <- ebg[chr4.idx]
library(GenomicAlignments)
se <- summarizeOverlaps(features = ebg.sub, reads = bf,
                        mode="Union",
                        singleEnd=FALSE,
                        ignore.strand=TRUE,
                        fragments=FALSE)
assays(se)$counts[1]

download.file("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/wang_eset.RData", "wang_eset.RData")
load("wang_eset.RData")
library(Biobase)
count.matrix <- exprs(wang.eset)[,10:21]
col.data <- pData(wang.eset)[10:21,]
library(DESeq2)
dds <- DESeqDataSetFromMatrix(count.matrix, col.data, design=~1)
sizeFactors(estimateSizeFactors(dds))
dds$cell.type
rld <- rlog(dds)
plotPCA(rld, intgroup=c("cell.type"))

rmeans <- rowMeans(assay(rld)) # row mean of rlog-transformed data
idx <- c(1,2,10,7,8,9,12) # pick some samples for visualization
mat <- assay(rld)[rmeans > 1,idx] # pull out a small matrix of rlog-transformed counts
colnames(mat) <- rld$cell.type[idx] # name the columns of matrix by cell type

panel.cor <- function(x, y, digits = 2, prefix = "", cex.cor, ...)  {
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- abs(cor(x, y))
  txt <- format(c(r, 0.123456789), digits = digits)[1]
  txt <- paste0(prefix, txt)
  if(missing(cex.cor)) cex.cor <- 0.8/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex.cor * r)
}
pairs(mat, asp=1, col=rgb(0,0,0,.3), lower.panel=panel.cor)

dds$type <- factor(ifelse(dds$cell.type %in% c("cerebellum","mixed.brain"), 
                          "brain", "other"))
dds$type <- relevel(dds$type, "other")
design(dds) <- ~ type
dds <- DESeq(dds)
res <- results(dds)

plotMA(res, ylim=c(-10,10))
row.names(res)[which.min(res@listData$pvalue)]

res2 <- results(dds, lfcThreshold=2)
plotMA(res2, ylim=c(-10,10))
summary(res2)

plotCounts(dds, which.min(res$padj), intgroup="type")
par(mfrow=c(3,3))
for (i in 1:9)  plotCounts(dds, order(res$padj)[i], intgroup="type")
top <- rownames(res)[head(order(res$stat, decreasing=TRUE), 20)]

library(org.Hs.eg.db)
keytypes(org.Hs.eg.db)
anno <- select(org.Hs.eg.db, keys=top,
               columns=c("SYMBOL","GENENAME"), 
               keytype="ENSEMBL")
anno[match(top, anno$ENSEMBL),]

download.file("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bottomly_eset.RData", "bottomly_eset.RData")
load("bottomly_eset.RData")
library(Biobase) 
count.matrix <- exprs(bottomly.eset)
col.data <- pData(bottomly.eset) 
library(DESeq2)
dds <- DESeqDataSetFromMatrix(count.matrix, col.data, design=~1)
dds$experiment.number <- factor(dds$experiment.number)

library(sva)
idx <- rowMeans(counts(dds)) > 1
dat <- counts(dds)[idx,]
mod <- model.matrix(~ strain, colData(dds))
mod0 <- model.matrix(~ 1, colData(dds))
svseq <- svaseq(dat, mod, mod0, n.sv=2)
plot(svseq$sv[,1], svseq$sv[,2], col=dds$experiment.number, pch=16)
legend("bottom", levels(dds$experiment.number), pch=16, col=1:3)
text(svseq$sv[,1], svseq$sv[,2], 1:ncol(dds), pos=1)

library("pasilla")
inDir = system.file("extdata", package="pasilla", mustWork=TRUE)        
countFiles = list.files(inDir, pattern="fb.txt$", full.names=TRUE)    
flattenedFile = list.files(inDir, pattern="gff$", full.names=TRUE) 

sampleTable = data.frame(row.names = c( "treated1", "treated2", "treated3","untreated1", "untreated2", "untreated3", "untreated4" ), condition = c("knockdown", "knockdown", "knockdown", "control", "control", "control", "control" ))                 

library("DEXSeq")  
dxd = DEXSeqDataSetFromHTSeq(countFiles, sampleData=sampleTable, 
  design= ~ sample + exon + condition:exon, flattenedfile=flattenedFile )
rmean = rowMeans(counts(dxd))  
dxd2L = dxd[seqnames(rowData(dxd)) == "chr2L" & rmean > 10,] 
dxd2L = dxd2L[1:1000,]  

dxd = estimateSizeFactors( dxd2L )
dxd = estimateDispersions( dxd )
dxd = testForDEU( dxd )
dxd = estimateExonFoldChanges( dxd, fitExpToVar="condition")
dxr = DEXSeqResults( dxd )
dxr@listData$groupID[which.min(dxr@listData$pvalue)]
dxr@listData$featureID[which.min(dxr@listData$pvalue)]

plotDEXSeq(dxr, 'FBgn0000256', norCounts=TRUE, displayTranscripts=TRUE)

library(cummeRbund)
myDir <- system.file("extdata", package="cummeRbund") 
gtfFile <- system.file("extdata/chr1_snippet.gtf",package="cummeRbund")
cuff <- readCufflinks(dir=myDir,gtfFile=gtfFile,genome="hg19",rebuild=TRUE)

gene.features <- annotation(genes(cuff))
head(gene.features)
isoforms.per.gene <- gene.features[!duplicated(gene.features$isoform_id),    c("gene_id","isoform_id")]
isoforms.per.gene <- isoforms.per.gene[order(isoforms.per.gene$isoform_id),]
head(isoforms.per.gene)

gene.tab <- table(isoforms.per.gene$gene_id)
sum(gene.tab == 1)

isoform.fpkm <- fpkm(isoforms(cuff))
head(isoform.fpkm)
table(isoform.fpkm$sample_name)
ips <- isoform.fpkm[isoform.fpkm$sample_name == "iPS",]
hesc <- isoform.fpkm[isoform.fpkm$sample_name == "hESC",]
stopifnot(all(ips$isoform_id == isoforms.per.gene$isoform_id))
stopifnot(all(hesc$isoform_id == isoforms.per.gene$isoform_id))  
ips.max <- sapply(split(ips$fpkm, isoforms.per.gene$gene_id), which.max)
head(ips.max)

hesc.max <- sapply(split(hesc$fpkm, isoforms.per.gene$gene_id), which.max)
head(hesc.max)

sum(ips.max[ips.max == hesc.max]>1)

ips1 <- ips.max[gene.tab > 1]
hesc1 <- hesc.max[gene.tab > 1]
mean(ips1==hesc1)


sum(table(isoforms.per.gene$gene_id)>1)
im2=ips.max[gene.tab[gene.tab>1]] 
hm2=hesc.max[gene.tab[gene.tab>1]]
res1=sapply(1:239, function(i)
match(im2[i],hm2[i],nomatch=0))
answer= sum(res1)/239

```

